flow360/component/simulation/validation/validation_context.py

```diff
@@ -0,0 +1,119 @@
+"""
+Module for validation context handling in the simulation component of Flow360.
+
+This module provides context management for validation levels and specialized field
+definitions for conditional validation scenarios.
+
+Features
+--------
+- This module allows for defining conditionally (context-based) required fields,
+  such as fields required only for specific scenarios like surface mesh or volume mesh.
+- It supports running validation only for specific scenarios (surface mesh, volume mesh, or case),
+  allowing for targeted validation flows based on the current context.
+- This module does NOT ignore validation errors; instead, it enriches errors with context
+  information, enabling downstream processes to filter and interpret errors based on scenario-specific requirements.
+"""
+
+import contextvars
+from typing import Optional
+
+from pydantic import Field
+
+SURFACE_MESH = "SurfaceMesh"
+VOLUME_MESH = "VolumeMesh"
+CASE = "Case"
+# when running validation with ALL, it will report errors happing in all scenarios in one validation pass
+ALL = "All"
+
+_validation_level_ctx = contextvars.ContextVar("validation_level", default=None)
+
+
+class ValidationLevelContext:
+    """
+    Context manager for setting the validation level.
+
+    Allows setting a specific validation level within a context, which influences
+    the conditional validation of fields based on the defined levels.
+    """
+
+    def __init__(self, level: str):
+        if level not in {None, SURFACE_MESH, VOLUME_MESH, CASE, ALL}:
+            raise ValueError(f"Invalid validation level: {level}")
+        self.level = level
+        self.token = None
+
+    def __enter__(self):
+        self.token = _validation_level_ctx.set(self.level)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        _validation_level_ctx.reset(self.token)
+
+
+def get_validation_level():
+    """
+    Retrieves the current validation level from the context.
+
+    Returns:
+        The current validation level, which can influence field validation behavior.
+    """
+    return _validation_level_ctx.get()
+
+
+# pylint: disable=invalid-name
+def ConditionalField(
+    default=None, *, relevant_for: Optional[str] = None, required: Optional[bool] = False, **kwargs
+):
+    """
+    Creates a field with conditional validation requirements.
+
+    Parameters
+    ----------
+    default : any, optional
+        The default value for the field.
+    relevant_for : str, optional
+        Specifies the scenario for which this field is relevant. The relevant_for is included in error->ctx
+    required : bool, optional
```

Comment:
It is not only for required. If something is not required, but failed, for example due to incorrect type, we will show error message with context information. Frontend can decide then what to do with this information: cannot press run if any "case" errors, but can do surface meshing

Reply:
Does this mean that all fields have to specify this context then?

Reply:
not all field, it can be inherited or empty. But since I defined this on all fields in SimulationParams class, all fields will have them now.

Reply:
During my testing I found that when a sub model (which does not have context) has validation error, the context will not show whether it is CASE/SM/VM related. Only the field that directly specified the "context" will add this its validation message. e.g.
```python
class SubModel(Flow360BaseModel):
    my_int:int

class ParentModel(Flow360BaseModel):
    sub_model: SubModel = validation_context.ContextField(context=validation_context.CASE)

def test_context_propagation():
    mm = ParentModel(sub_model=SubModel(my_int="a"))
```
results in:
```
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for SubModel
E       my_int
E         Input should be a valid integer, unable to parse string as an integer [type=int_parsing, input_value='a', input_type=str]
E           For further information visit https://errors.pydantic.dev/2.7/v/int_parsing
```

which does not have context. So I would imagine we need to replace almost all `pd.Field` with `ContextField` or `ConditionalField`. But that is not the main purpose of this PR anyway.

Reply:
This is correct. Your code will not propagate context because it fails on executing `SubModel(my_int="a")` and is not reaching out to `ParentModel`. But if the data is provided as a dictionary, the error message will be created correctly:

```python
from flow360.component.simulation.framework.base_model import Flow360BaseModel
from flow360.component.simulation.validation import validation_context
import pydantic as pd

class SubModel(Flow360BaseModel):
    my_int:int

class ParentModel(Flow360BaseModel):
    sub_model: SubModel = validation_context.ContextField(context=validation_context.CASE)

def test_context_propagation():
    try:
        mm = ParentModel(sub_model=SubModel(my_int="a"))
    except Exception as e:
        print(e.errors())

def test_context_propagation_dict():
    try:
        mm = ParentModel(**dict(sub_model=dict(my_int="a")))
    except Exception as e:
        print(e.errors())

test_context_propagation()
test_context_propagation_dict()
```
[{'type': 'int_parsing', 'loc': ('my_int',), 'msg': 'Input should be a valid integer, unable to parse string as an integer', 'input': 'a', 'url': 'https://errors.pydantic.dev/2.7/v/int_parsing'}]
[{'type': 'int_parsing', 'loc': ('sub_model', 'my_int'), 'msg': 'Input should be a valid integer, unable to parse string as an integer', 'input': 'a', 'ctx': {'relevant_for': 'Case'}, 'url': 'https://errors.pydantic.dev/2.7/v/int_parsing'}]
```

Note that webUI will always see the "dict" version.

Not sure if this is expected behaviour or not? Probably it is expected since we want to strictly validate whatever user explicitly input. A different thing is on submit, since we can ignore some of the errors. On submit we can do context validation of entire model, so we will have context in all error messages. The only issue now is that we are not able to create valid SimulationParams object if any errors are present regardless of context.

Reply:
I think as long as we can have the full and properly "contexted" validation I am fine. The `ContextField` is more of a prompt/nice to have.